### DISCUSSION TITLE
Can't display footnotes in MDX (gatsby-plugin-mdx@^5.x.x)

### DISCUSSION THREAD

**Discussion by code-istvan** [gatsbyjs/gatsby#37955] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Hi my name is Istvan!

I'm trying to display a footnote using gatsby-plugin-mdx@^5.x.x and the `gatsby-transformer-remark` and `gatsby-remark-footnes` but since the upgrade of gatsby-plugin-mdx@^5.x.x I'm not able to do it. I don't have a body to fetch and I don't know how to display them.

This is how I'm rendering it in the article template:
```js
<article className="mb-20px blog-posts-body-style">
     {children}
</article>
```
This is my query:
```
const query = graphql`
  query PostsBySlug($slug: String!) {
    mdx(fields: { slug: { eq: $slug } }) {
      id
      excerpt
      fields {
        slug
        timeToRead {
          minutes
        }
      }
      frontmatter {
        author
        title
        date(formatString: "YYYY MMMM DD.", locale: "hu-HU")
        tags
        thumbnail
      }
    }
  }
`

```

This is my gatsby-config:
```
{
      resolve: `gatsby-transformer-remark`,
      options: {
        plugins: [
          {
            resolve: `gatsby-remark-footnotes`,
            options: {
              footnoteBackRefPreviousElementDisplay: "inline",
              footnoteBackRefDisplay: "inline",
              footnoteBackRefInnerText: "^",
              footnoteListTag: "div",
              footnoteBackRefAnchorStyle: `text-decoration: none;`,
              footnoteBackRefInnerTextStartPosition: "front",
              useFootnoteMarkerText: false,
            },
          },
        ],
      },
    },

```

### Reproduction Link

https://github.com/code-istvan/bw

### Steps to Reproduce

1. Clone the repo
2. Install dep. NPM
3. Run gatsby develop
4. Go to the http://localhost:8000/blog/test/

### Expected Result

I would like to display the footnotes.

### Actual Result

The footnote rendered as a plain text

<img width="440" alt="Képernyőfotó 2023-04-17 - 18 02 48" src="https://user-images.githubusercontent.com/63305129/232543920-223c5616-f210-4669-8fea-8325bdbdcbd9.png">

### Environment

```shell
System:
    OS: macOS 13.2.1
    CPU: (8) arm64 Apple M1 Pro
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.9.0 - ~/.nvm/versions/node/v18.9.0/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 8.19.1 - ~/.nvm/versions/node/v18.9.0/bin/npm
  Browsers:
    Chrome: 112.0.5615.49
    Firefox: 111.0.1
    Safari: 16.3
  npmPackages:
    gatsby: ^5.8.1 => 5.8.1 
    gatsby-core-utils: ^4.8.0 => 4.8.0 
    gatsby-plugin-breakpoints: ^1.3.9 => 1.3.9 
    gatsby-plugin-gdpr-cookies: ^2.0.9 => 2.0.9 
    gatsby-plugin-image: ^3.8.0 => 3.8.0 
    gatsby-plugin-loadable-components-ssr: ^4.3.2 => 4.3.2 
    gatsby-plugin-manifest: ^5.8.0 => 5.8.0 
    gatsby-plugin-mdx: ^5.8.0 => 5.8.0 
    gatsby-plugin-netlify: ^5.1.0 => 5.1.0 
    gatsby-plugin-netlify-cms: ^7.8.0 => 7.8.0 
    gatsby-plugin-nprogress: ^5.8.0 => 5.8.0 
    gatsby-plugin-offline: ^6.8.0 => 6.8.0 
    gatsby-plugin-page-creator: ^5.8.0 => 5.8.0 
    gatsby-plugin-robots-txt: ^1.8.0 => 1.8.0 
    gatsby-plugin-sass: ^6.8.0 => 6.8.0 
    gatsby-plugin-sharp: ^5.8.1 => 5.8.1 
    gatsby-plugin-sitemap: ^6.8.0 => 6.8.0 
    gatsby-remark-footnotes: ^0.0.8 => 0.0.8 
    gatsby-remark-images: ^7.8.0 => 7.8.0 
    gatsby-remark-numbered-footnotes: ^1.0.1 => 1.0.1 
    gatsby-remark-relative-images: ^2.0.2 => 2.0.2 
    gatsby-remark-relative-images-v2: ^0.1.5 => 0.1.5 
    gatsby-source-filesystem: ^5.8.0 => 5.8.0 
    gatsby-transformer-json: ^5.8.0 => 5.8.0 
    gatsby-transformer-remark: ^6.8.0 => 6.8.0 
    gatsby-transformer-sharp: ^5.8.0 => 5.8.0 
  npmGlobalPackages:
    gatsby-cli: 5.2.0
```

**Comment by LekoArts** [comment, marked as answer]
You need to install https://github.com/remarkjs/remark-gfm and use it inside the `remarkPlugins` options of `gatsby-plugin-mdx`: https://www.gatsbyjs.com/plugins/gatsby-plugin-mdx/#mdxoptions

Uninstalll `gatsby-transformer-remark` and `gatsby-remark-footnes`, you don't need them and they do nothing for your MDX files.

**Reply by fbuireu** [reply]
Hi! Sorry to jump in but I'm facing the same issue and I also tried @LekoArts suggestion. 

I've cloned [@code-istvan's project ](https://github.com/code-istvan/bw)and after cleaning the dependencies I added the following `gatsby-config.js` configuration.

### Using `remark-gfm@^1`:

```js
    {
      resolve: `gatsby-plugin-mdx`,
      options: {
        extensions: [`.mdx`, `.md`],
        mdxOptions:{
          remarkPlugins: [
              wrapESMPlugin(`remark-gfm`)
          ],
        },
      },
    },
```

The compilation crashes with the following error message:

```
Failed to compile the file ".../bw/blog/a-legzes-legtokeletesebb-modja-az-astanga-jogaban.md". Original error message:

this.data is not a function
```

### Using `remark-gfm@latest`:

```js
    {
      resolve: `gatsby-plugin-mdx`,
      options: {
        extensions: [`.mdx`, `.md`],
        mdxOptions:{
          remarkPlugins: [
            require(`remark-gfm`),
          ],
        },
      },
    },
```
The code breaks because of ES Modules. 

 ```
Error: [ERR_REQUIRE_ESM]: require() 
 Instead change the require of index.js in bw/gatsby-config.js to a dynamic import() which is available in all CommonJS modules.
```

Using a dynamic import approach outputs the same error as: `remark-gfm@^1`

### Using @mdx-js/mdx@v1 and @mdx-js/react@v1 as https://github.com/gatsbyjs/gatsby/discussions/34714#discussioncomment-2108962 

Not sure how to address this configuration along with the previous `remarkPlugins` ES Modules/CommonJS but the same issue still as before (can compile with `remark-gfm@^1` but without displaying footnotes and ES Module error with latest)

### Switching to `gatsby-config.mjs`
As [`v5.3` release notes shows](https://www.gatsbyjs.com/docs/reference/release-notes/v5.3/), and after doing all the mandatory changes (`__dirname`, etc), used as:

```js
import remarkGfm from 'remark-gfm'


    {
      resolve: `gatsby-plugin-mdx`,
      options: {
        extensions: [`.mdx`, `.md`],
        mdxOptions:{
          remarkPlugins: [
            remarkGfm
          ],
        },
      },
    },
```
The code compiles yet the footnotes are still not present.

I guess that all the approaches that doesn't break the code (hence compiles) are not displaying the footnotes because of the way the article is being rendered (and because the removal of the `MDXProvider`, rendered as a `children`) not as content itself.

What's the approach to follow in those cases?

Let me know if you need further details on my side

_P.S: the project uses Node 18 in case anyone wants to clone it._

**Reply by LekoArts** [reply]
You should use `gatsby-config.mjs` as the updated README shows (this will be released today): https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-mdx

Here's a working example: https://minimal-blog.lekoarts.de/markdown-reference
Config: https://github.com/LekoArts/gatsby-themes/blob/main/themes/gatsby-theme-minimal-blog-core/gatsby-config.mjs
Source file: https://github.com/LekoArts/gatsby-themes/blob/main/examples/minimal-blog/content/posts/markdown-reference/index.mdx

**Reply by fbuireu** [reply]
Thanks @LekoArts! It worked when I updated `remark-gfm` to its latest version + using `.mjs` config file!

**Reply by LekoArts** [reply]
Yeah, older versions of `remark-gfm` didn't ship with footnotes support, see https://github.com/remarkjs/remark-gfm/releases/tag/3.0.0

**Reply by code-istvan** [reply]
Thank ou so much for your help @LekoArts and @fbuireu! After the above changes the footnote works perfectly for me. Have a nice day with clean code!

**Reply by LekoArts** [reply]
Glad to hear you all got it working 👍
